src/Microsoft.AspNetCore.Blazor/Rendering/RenderBatchBuilder.cs

```diff
@@ -27,6 +27,12 @@ internal class RenderBatchBuilder
         public Queue<RenderQueueEntry> ComponentRenderQueue { get; } = new Queue<RenderQueueEntry>();
         public Queue<int> ComponentDisposalQueue { get; } = new Queue<int>();
 
+        // Scratch data structure for understanding attribute diffs.
+        //
+        // One of these can be used for the whole tree diff because attributes cannot be nested.
+        // Does not need to be cleared because it is always emptied at the end of each attribute pass.
```

Comment:
What do you think about the risk that on an earlier pass, it didn't get emptied due to being halted by an exception? Should we say that behavior is undefined from that point on, or would it be preferable to empty the dict at the start, since it should be nearly free anyway if it's already empty?

Reply:
That's an interesting thought. Since we don't expect anything during the diff pass to *handle* the exception, I propose that we put a 'reset' at the top level in a try/finally. Then it's a once-per-pass cost.